make pre-commit use venv's even for `meta` hooks

MatrixManAtYrService opened the pull request:
I'm not a seasoned nixpkgs contributor, nor a pre-commit expert, so I'm hoping somebody comes along and shows me a better way.  This is a little hacky.

### The Problem

Previously, the nixpkgs pre-commit package propagated its build inputs into user devshells.  This caused [problems](https://github.com/nix-community/poetry2nix/issues/1076), so I "[fixed it](https://github.com/NixOS/nixpkgs/pull/235123)" to make pre-commit more self-contained.  

That is, to make it behave more like an app and less like a library.  I made it so that you can't import pre-commit or its dependencies in your python code without explicitly including it in your python environment.  i.e. having it as a package in your nix shell lets you run it like an app, but not import it in your code.   

This fix was premature.  It turns out that [I broke meta imports](https://github.com/NixOS/nixpkgs/issues/238345)(shame on me for not testing more thoroughly).

Most pre-commit hooks live in external repositories, so pre-commit clones them and creates environments for them.  `meta` hooks, however live in the pre-commit repo and are called directly.  So unlike the others, they're expecting pre-commit to be treated "like a library" and not "like an app".

When I switched the package from library mode to app mode, I removed pre-commit from the environment that the `meta` hooks were using.  Now they break like so:

```
Check hooks apply to the repository......................................Failed
- hook id: check-hooks-apply
- exit code: 1

/nix/store/6jx4nc20v3cfphlbypqid30q41i5vh5x-python3-3.10.11/bin/python3.10: Error while finding module specification for 'pre_commit.meta_hooks.check_hooks_apply' (ModuleNotFoundError: No module named 'pre_commit')
```

I'm uncertain about whether this is better or worse than it was before.  On one hand, the failure is more likely to appear.  On the other hand, it's less of a mystery and easier to work around.  These hooks are linting your pre-commit config so that *other* hooks can lint your code (or whatever else you're using pre-commit for).  They're more safely ignored than some others.  It's still no fun that they break.

### My Solution

**EDIT: This solution was abandoned, scroll down to @anund's comments to see the better one**

The patch included in this PR mutates the pre-commit repo so that it provides the `meta` hooks in the same way that any other repo might provide any other hook (see comments below for how this works).

MatrixManAtYrService commented on pkgs/tools/misc/pre-commit/meta-hook-rewiring.patch:
```diff
@@ -0,0 +1,182 @@
++
++local_repo = {'repo': REPO_ROOT, 'rev': 'HEAD'}
++
++def replace_meta(config_in: Config) -> Config:
```
Typically, a user can just use the string "meta" to indicate a meta hook.  That string is then used by pre-commit to avoid certain steps that would be necessary for an external hook.

Here I modify the way pre-commit reads config from disk so that "meta" is replaced by the contents of the `REPO_ROOT` variable.  I've tested this with `REPO_ROOT` set to `/home/matt/src/pre-commit` (where I had cloned pre-commit by hand) and as long as that repo has these changes, it works.

The goal is to provide a derivation which includes these changes in `/nix/store` so that pre-commit always sees meta hooks as external repositories.  Except "external" in this case just means that they're provided by the pre-commit repo that lives in the nix store.

MatrixManAtYrService commented on pkgs/tools/misc/pre-commit/default.nix:
```diff
@@ -70,6 +97,14 @@ buildPythonApplication rec {
+    # meta-hook-rewiring needs a git repo, not just a source dir
+    # let it have one in the nix store
+    echo SRC ${src}
+    substituteInPlace pre_commit/clientlib.py \
```
This substitution is supposed to populate the REPO_ROOT variable, but as far as I can tell it doesn't work either

anund replied:
Superficially it looks like you're missing surrounding @'s in the patch. See https://nixos.org/manual/nixpkgs/stable/#fun-substitute.

MatrixManAtYrService commented on pkgs/tools/misc/pre-commit/default.nix:
```diff
@@ -166,6 +201,19 @@ buildPythonApplication rec {
     # but we use an absolute path so it's not an issue.
     "test_environment_not_sourced"
+
+    # todo: explanations for these
```
some of these failures make sense.  I've added two hooks to the repo and they're assuming that I haven't.  I am not yet sure that they can *all* be safely ignored.

MatrixManAtYrService commented:
I'll be on vacation for a little while, so I may not be working on this for the next few weeks.  I'd love to come back to find that somebody has a smarter way forward, but I wanted to post this as a draft in the meantime in case that ends up being helpful.

I feel bad for having broken this, but maybe reverting it to its previous (also broken state) is just as bad?

MatrixManAtYrService commented:
I think this might be overkill.  I don't need to convince pre-commit to build a separate venv  that includes... itself.  I just need to hack the path so that it knows about the non-venv instance of itself when it needs to import those internal checks.  I'm going to play around with that idea before continuing with this one.

anund commented:
Poking around a little after seeing this mentioned on discourse:
- at first glance your previous change should have been equivalent to changing `propogatedBuildInputs` to `buildInputs`. Though you'd still end up with a python reference being passed https://github.com/NixOS/nixpkgs/blob/31b1973dc4eead85949acbfcbab0bf010de168a2/pkgs/development/interpreters/python/mk-python-derivation.nix#L207-L212
- changing to `buildInputs` wouldn't work as `buildPythonApplication` relies on `nix-support/propagated-build-inputs`  to build an import path https://github.com/NixOS/nixpkgs/blob/31b1973dc4eead85949acbfcbab0bf010de168a2/pkgs/development/interpreters/python/wrap.sh#L115-L122
- pre-commit is using ~`sys.executable -m precommit....` to launch meta hooks https://github.com/pre-commit/pre-commit/blob/23df082edcf80f8501d94e8a71e237d1e69ffb15/pre_commit/clientlib.py#L212

It's unclear to me just when `PYTHON_PATH` is contaminated as part of the `mkShell` call though that can be ignored for now.

I wonder if the files in https://github.com/pre-commit/pre-commit/tree/23df082edcf80f8501d94e8a71e237d1e69ffb15/pre_commit/meta_hooks can also be wrapped the way the main function is.

anund commented:
Getting a reproducible error case by adding this file to the root of nixpkgs.
```yaml
# .pre-commit-config.yaml
repos:
  - repo: meta
    hooks:
      - id: identity
```


```bash
$ nix-build -A pre-commit
$ result/bin/pre-commit install #probably?
$ result/bin/pre-commit install-hooks #probably?
$ result/bin/pre-commit run --all-files
# see error loading identity
```

Looking at the output wrapped program I can see the first level wrapper isn't setting `PYTHON_PATH`. It does set up and exports `PATH` and `PYTHONNOUSERSITE`. (`$cat result/bin/precomit`) The wrapped/exec'd `.pre-commit-wrapped` has an import statement that sets up all the needed imports to run the program. (which is why it runs)

Since the meta sub modules are just being invoked with, effectively, `python -m pre-comit....` not setting PYTHON_PATH in the wrapping process means the `-m` call is missing all the import setup in the wrapped import statments. (see https://docs.python.org/3/using/cmdline.html#environment-variables for other envvars python looks for during invocation)

The good news is there's a simple fix to just have the wrapper set `PYTHON_PATH` via https://github.com/NixOS/nixpkgs/blob/31b1973dc4eead85949acbfcbab0bf010de168a2/pkgs/development/interpreters/python/mk-python-derivation.nix#L64-L66 (not this can be `--set or --prefix` whichever makes more sense for the expected behaviour)

```diff
diff --git a/pkgs/tools/misc/pre-commit/default.nix b/pkgs/tools/misc/pre-commit/default.nix
index ded7b3290d3c..b6fe67715cce 100644
--- a/pkgs/tools/misc/pre-commit/default.nix
+++ b/pkgs/tools/misc/pre-commit/default.nix
@@ -101,6 +101,8 @@ buildPythonApplication rec {
     deactivate
   '';
 
+  makeWrapperArgs = ''--set PYTHONPATH $PYTHONPATH'';
+
   # Propagating dependencies leaks them through $PYTHONPATH which causes issues
   # when used in nix-shell.
   postFixup = ''
```

anund commented:
ping @MatrixManAtYrService the above might help

MatrixManAtYrService commented:
@anund I'll be taking a closer look in a day or two. Thanks very much for this.

MatrixManAtYrService committed:
pre-commit: fixing import bug for built-in hooks

In a previous change, I prevented PYTHONPATH from propatating into nix
devshells.  This fixed some problems where pre-commit's dependenceis
were being used instead of the user's dependencies.

But pre-commit bundles three pre-commit hooks internally (as opposed to
"normal" hooks, which are references to external repo's).  By removing
PYTHONPATH, I prevented these internal hooks from being correctly
imported.

This change adds PYTHONPATH to the wrapper which launches pre-commit so
that those built-in hooks can be called directly.

MatrixManAtYrService commented:
@anund As far as I can tell, adding PYTHONPATH in `makeWrapperArgs` does the trick, and is a much better solution than the direction I was pursuing.  Thanks again.

I have some pre-commit-related work to do.  I'm going to use this change in that work as a way of testing it a bit more thoroughly, and if I still think it scratches the itch after a few days, I'll mark it ready for review.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/where-did-you-get-stuck-in-the-nix-ecosystem-tell-me-your-story/31415/60

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/where-did-you-get-stuck-in-the-nix-ecosystem-tell-me-your-story/31415/62

MatrixManAtYrService commented:
I've been testing this and I think it's good. I don't want reviewers to have to wade through the saga above, however, so I'm going to close it and make a nice clean PR.
